### PR TITLE
Add read/write lock in the fault handler

### PR DESCRIPTION
**Pull request by xxx0624**
### Summary

This PR is to add a field / RWMutex in the fault handler which can help to avoid race condition where multiple clients would manipulate the same network resource at the same time.

### Implementation details

- Add the read/write lock in the [FaultHandler](https://github.com/aws/amazon-ecs-agent/blob/dev/ecs-agent/tmds/handlers/fault/v1/handlers/handlers.go#L41C6-L41C18)
- Add a new method to create the [handler](https://github.com/aws/amazon-ecs-agent/blob/dev/ecs-agent/tmds/handlers/fault/v1/handlers/handlers.go#L41C6-L41C18) for faults
- Add more validations in [validateTaskNetworkConfig](https://github.com/aws/amazon-ecs-agent/blob/5589a96639d216db755ddde6fb45764376889a84/ecs-agent/tmds/handlers/fault/v1/handlers/handlers.go#L704-L712) to make sure the required field has non-empty value.

### Testing

New tests cover the changes: 
No. Existing unit test will cover this.

### Description for the changelog

Feature: Add RWMutex in the fault handler.

### Additional Information

**Does this PR include breaking model changes? If so, Have you added transformation functions?**
  

**Does this PR include the addition of new environment variables in the README?**

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: ecs-agent/tmds/handlers/fault/v1/handlers/handlers_test.go [hunk 1]
```diff
@@ -264,6 +297,84 @@ func generateNetworkBlackHolePortTestCases(name string, expectedHappyResponseBod
...
+				fmt.Sprintf("failed to get task metadata due to internal server error for container: %s", endpointId)),
+			setAgentStateExpectations: func(agentState *mock_state.MockAgentState) {
+				agentState.EXPECT().GetTaskMetadata(endpointId).Return(state.TaskResponse{
+					TaskResponse:          &v2.TaskResponse{TaskARN: taskARN},
+					FaultInjectionEnabled: true,
+					TaskNetworkConfig: &state.TaskNetworkConfig{
+						NetworkMode:       awsvpcNetworkMode,
+						NetworkNamespaces: nil,
+					},
+				}, nil)
```
Comment (mye956) [inline ecs-agent/tmds/handlers/fault/v1/handlers/handlers_test.go:314]: nit (non-blocking): might be a good idea to add a `Times(1)` for these test cases as well if want to make things consisent. wdyt?

**Review by mye956 — APPROVED** [review]

File: ecs-agent/tmds/handlers/fault/v1/handlers/handlers_test.go (hunk 1 above)
Comment (xxx0624) [reply ecs-agent/tmds/handlers/fault/v1/handlers/handlers_test.go:314]: Sounds good. I can add it in a follow up change if no major concern in this PR.

File: ecs-agent/tmds/handlers/fault/v1/handlers/handlers.go [hunk 2]
```diff
@@ -206,12 +238,21 @@ func (h *FaultHandler) StartNetworkLatency() func(http.ResponseWriter, *http.Req
...
 		if err != nil {
 			return
 		}
 
+		// the network latency fault would be injected to given elastic network
+		// interface/ENI.
+		eniID := taskMetadata.TaskNetworkConfig.
+			NetworkNamespaces[0].
+			NetworkInterfaces[0].
+			ENIID
```
Comment (amogh09) [inline ecs-agent/tmds/handlers/fault/v1/handlers/handlers.go:251]: Why not use netns path here also?

Comment (xxx0624) [reply ecs-agent/tmds/handlers/fault/v1/handlers/handlers.go]: Updated.

File: ecs-agent/tmds/handlers/fault/v1/handlers/handlers.go [hunk 3]
```diff
@@ -42,18 +43,36 @@ const (
 )
 
 type FaultHandler struct {
-	// TODO: Mutex will be used in a future PR
-	// mu             sync.Mutex
+	// mutexMap is used to avoid multiple clients to manipulate same resource at same
+	// time. The 'key' is the ENI ID or the network namespace path and 'value' is the
```
[the remark this answers is not in the document]
Comment (xxx0624) [reply ecs-agent/tmds/handlers/fault/v1/handlers/handlers.go]: Updated.

**Review by mye956 — APPROVED** [review]

File: agent/handlers/v4/tmdsstate.go [hunk 4]
```diff
@@ -161,7 +161,7 @@ func (s *TMDSAgentState) getTaskMetadata(v3EndpointID string, includeTags bool)
 					Path: task.GetNetworkNamespace(),
 					NetworkInterfaces: []*tmdsv4.NetworkInterface{
 						{
-							DeviceName: "",
+							DeviceName: "ethx",
```
Comment (amogh09) [inline agent/handlers/v4/tmdsstate.go:164]: Why this change?

Comment (amogh09) [reply agent/handlers/v4/tmdsstate.go:164]: Thanks!

Comment (xxx0624) [reply agent/handlers/v4/tmdsstate.go:164]: Some context:
The `DeviceName` is required for network latency / network packet loss faults. We are exposing this information (and other task network config) via AgentState to facilitate the fault injection handler to start/stop/check network fault.

Why we need a non-empty value here which is fake: because in this PR, we introduce a new check ([link](https://github.com/aws/amazon-ecs-agent/blob/02cb57659a882844033d0c18c7e666609347fb61/ecs-agent/tmds/handlers/fault/v1/handlers/handlers.go#L688-L691)) to see if the deviceName is empty. So if that's empty, these [unit test](https://github.com/aws/amazon-ecs-agent/blob/78a2bf0c7d3ebd3a13de3ac733af46dfb3816b18/agent/handlers/task_server_setup_test.go#L3695-L3708) will fail.

File: ecs-agent/tmds/handlers/fault/v1/handlers/handlers.go [hunk 5]
```diff
@@ -416,11 +473,17 @@ func (h *FaultHandler) CheckNetworkPacketLoss() func(http.ResponseWriter, *http.
 
 		// Obtain the task metadata via the endpoint container ID
 		// TODO: Will be using the returned task metadata in a future PR
-		_, err = validateTaskMetadata(w, h.AgentState, requestType, r)
+		taskMetadata, err := validateTaskMetadata(w, h.AgentState, requestType, r)
```
Comment (tshan2001) [inline ecs-agent/tmds/handlers/fault/v1/handlers/handlers.go:476]: Do we need this in all check methods? Since we're only calling `tc q` and check if `"kind":"netem"` exists, we don't need to know the network namespace path

Comment (xxx0624) [reply ecs-agent/tmds/handlers/fault/v1/handlers/handlers.go:476]: for awsvpc mode, all commands should be running in the given task network namespace. So we also need this check here.

**Review by amogh09 — APPROVED** [review]

**Review by tshan2001 — APPROVED** [review]

**Review by mye956 — APPROVED** [review]
